packages/headless/doc-parser/src/extractor.ts

```diff
@@ -25,58 +26,83 @@ enum Level {
   Type,
 }
 
-export function extractTypes(headingEntities: AnyEntity[]): Extraction {
+export function extractTypes(
+  headingEntities: AnyEntity[],
+  referencesCount: InterfaceReferencesCount
+): Extraction {
   const extraction: Extraction = {
     types: [],
   };
 
-  processObjectEntities(headingEntities, extraction, Level.Heading);
-  processFunctionEntities(headingEntities, extraction);
+  processObjectEntities(
+    headingEntities,
+    extraction,
+    Level.Heading,
+    referencesCount
+  );
+  processFunctionEntities(headingEntities, extraction, referencesCount);
 
   return extraction;
 }
 
 function processObjectEntities(
   entities: AnyEntity[],
   extraction: Extraction,
-  level: Level
+  level: Level,
+  referencesCount: InterfaceReferencesCount
 ) {
   entities
     .filter(isObjectEntity)
-    .forEach((entity) => extract(entity, extraction, level + 1));
+    .forEach((entity) =>
+      extract(entity, extraction, level + 1, referencesCount)
+    );
 }
 
 function processFunctionEntities(
   entities: AnyEntity[],
-  extraction: Extraction
+  extraction: Extraction,
+  referencesCount: InterfaceReferencesCount
 ) {
   entities.filter(isFunctionEntity).forEach((func) => {
     const {params, returnType} = func;
     const entities = isString(returnType) ? params : [...params, returnType];
 
-    processObjectEntities(entities, extraction, Level.Key);
+    processObjectEntities(entities, extraction, Level.Key, referencesCount);
   });
 }
 
-function extract(entity: ObjEntity, extraction: Extraction, level: Level) {
+function extract(
+  entity: ObjEntity,
+  extraction: Extraction,
+  level: Level,
+  referencesCount: InterfaceReferencesCount
+) {
   const {members} = entity;
 
   if (!members.length) {
     return;
   }
 
-  if (level === Level.Type) {
+  if (
+    !entity.isTypeExtracted &&
+    (level === Level.Type || referencesCount[entity.typeName] > 1)
+  ) {
```

Comment:
Prior to this PR, it was impossible for `isTypeExtracted` to be `true` during the extraction phase.

Now, when an interface is referenced more than once, only the first reference will have `isTypeExtracted` set to `false`. All subsequently parsed interfaces (where the children weren't parsed) are going to have `isTypeExtracted` set to `true`. This condition will move the first referenced interface (whose children are pased) to the `extractedTypes` section of the final JSON.